src/main/drivers/pitotmeter_ms4525.c

```diff
@@ -19,86 +19,125 @@
-static uint16_t ms4525_up;  // static result of pressure measurement
-static uint8_t rxbuf[4];
+typedef struct __attribute__ ((__packed__)) ms4525Ctx_s {
+    bool     dataValid;
+    uint32_t ms4525_ut;
+    uint32_t ms4525_up;
+} ms4525Ctx_t;
+
+STATIC_ASSERT(sizeof(ms4525Ctx_t) < BUS_SCRATCHPAD_MEMORY_SIZE, busDevice_scratchpad_memory_too_small);
 
-static void ms4525_start(pitotDev_t * pitot)
+static bool ms4525_start(pitotDev_t * pitot)
 {
-    busReadBuf( pitot->busDev, 0xFF, rxbuf, 4 );
+    uint8_t rxbuf[1];
+    bool ack = busReadBuf(pitot->busDev, 0xFF, rxbuf, 1);
+    return ack;
 }
 
-static void ms4525_read(pitotDev_t * pitot)
+static bool ms4525_read(pitotDev_t * pitot)
 {
-    if (busReadBuf( pitot->busDev, 0xFF, rxbuf, 4 )) {
-        ms4525_up = (rxbuf[0] << 8) | (rxbuf[1] << 0);
-        ms4525_ut = ((rxbuf[2] << 8) | (rxbuf[3] << 0))>>5;
+    uint8_t rxbuf1[4];
+    uint8_t rxbuf2[4];
+
+    ms4525Ctx_t * ctx = busDeviceGetScratchpadMemory(pitot->busDev);
+    ctx->dataValid = false;
+
+    if (!busReadBuf(pitot->busDev, 0xFF, rxbuf1, 4)) {
+        return false;
     }
+
+    if (!busReadBuf(pitot->busDev, 0xFF, rxbuf2, 4)) {
+        return false;
+    }
+
+    const uint8_t status = ((rxbuf1[0] & 0xC0) >> 6);
+    if (status == 2 || status == 3) {
+        return false;
+    }
+
+    int16_t dP_raw1, dT_raw1;
+    int16_t dP_raw2, dT_raw2;
+
+    dP_raw1 = 0x3FFF & ((rxbuf1[0] << 8) + rxbuf1[1]);
+    dT_raw1 = (0xFFE0 & ((rxbuf1[2] << 8) + rxbuf1[3])) >> 5;
+    dP_raw2 = 0x3FFF & ((rxbuf2[0] << 8) + rxbuf2[1]);
+    dT_raw2 = (0xFFE0 & ((rxbuf2[2] << 8) + rxbuf2[3])) >> 5;
+
+    // reject any double reads where the value has shifted in the upper more than 0xFF
+    if (ABS(dP_raw1 - dP_raw2) > 0xFF || ABS(dT_raw1 - dT_raw2) > 0xFF) {
+        return false;
+    }
+
+    // Data valid, update ut/up values
+    ctx->dataValid = true;
+    ctx->ms4525_up = (dP_raw1 + dP_raw2) / 2;
+    ctx->ms4525_ut = (dT_raw1 + dT_raw2) / 2;
+    return true;
 }
 
 static void ms4525_calculate(pitotDev_t * pitot, float *pressure, float *temperature)
 {
-    UNUSED(pitot);
-
-    uint8_t status = (ms4525_up & 0xC000) >> 14;
-    switch (status) {
-        case 0:
-            break;
-        case 1:
-            /* fallthrough */
-        case 2:
-            /* fallthrough */
-        case 3:
-            return;
-    }
-    int16_t dp_raw = 0, dT_raw = 0;
+    ms4525Ctx_t * ctx = busDeviceGetScratchpadMemory(pitot->busDev);
 
-    /* mask the used bits */
-    dp_raw = 0x3FFF & ms4525_up;
-    dT_raw = ms4525_ut;
+    const float P_max = 1.0f;
+    const float P_min = -P_max;
+    const float PSI_to_Pa = 6894.757f;
 
-    float dP = (10 * (int32_t)(dp_raw)) * 0.1052120688f;
-    float T  = (float)(200 * (int32_t)dT_raw - 102350) / 2047 + 273.15f;
+    //float dP = ctx->ms4525_up * 10.0f * 0.1052120688f;
+    const float dP_psi = -((ctx->ms4525_up - 0.1f * 16383) * (P_max - P_min) / (0.8f * 16383) + P_min);
```

Comment:
Is the Negative Sign('-') Intentional? Seems redundant anyway, but I wondered if that had any physical meaning :/